### PR TITLE
add capability for reverse bias

### PR DESCRIPTION
**Pull request by cwhanse**
- [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries to [`docs/sphinx/source/api.rst`](https://github.com/pvlib/pvlib-python/blob/master/docs/sphinx/source/api.rst) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/master/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

Adds reverse bias capability to single diode functions. The Lambert W technique cannot solve the single diode equation with the term for current at reverse bias. The Bishop '88 methods can solve the equation. Tested with `method='newton'`, not clear if `method='brentq'` is reliable for this application.

### CODE REVIEW

**Comment by cwhanse** [comment]
Ready for review.  

stickler complains about breaks both before and after an operator. Suggestions?  Test failures are for `test_forecast.py`.

File: pvlib/tests/test_singlediode.py [hunk 1]
```diff
@@ -185,3 +185,103 @@ def test_pvsyst_recombination_loss(method, poa, temp_cell, expected, tol):
...
+                'voc': 79.646
+            },
+            (1e-4, 1e-4)
+        )
+    ]
+)
+@pytest.mark.parametrize('method', ['newton', 'brentq'])
+def test_pvsyst_breakdown(method, brk_params, recomb_params, poa, temp_cell,
+                          expected, tol):
+    """test PVSst recombination loss"""
```
Comment (CameronTStark) [inline pvlib/tests/test_singlediode.py:226]:
```suggestion
    """test PVSyst recombination loss"""
```
Perhaps remove/change to align with test name.

File: pvlib/singlediode.py [hunk 2]
```diff
@@ -176,7 +217,9 @@ def bishop88(diode_voltage, photocurrent, saturation_current,
 
 def bishop88_i_from_v(voltage, photocurrent, saturation_current,
                       resistance_series, resistance_shunt, nNsVth,
-                      d2mutau=0, NsVbi=np.Inf, method='newton'):
+                      d2mutau=0, NsVbi=np.Inf, breakdown_factor=0.,
+                      breakdown_voltage=-5.5, breakdown_exp=3.28,
+                      method='newton'):
```
Comment (CameronTStark) [inline pvlib/singlediode.py:222]: As a convention, is it better to add new kwargs at the end, E.g. after `method='newton'`, so it doesn't break code that relies on position than explicit `kwarg=` usage? 

I'm half asking if that's something that's been historically done.

File: pvlib/singlediode.py [hunk 3]
```diff
@@ -150,21 +173,39 @@ def bishop88(diode_voltage, photocurrent, saturation_current,
...
         g_diode = saturation_current * np.exp(v_star) / nNsVth  # conductance
-        grad_i = -g_diode - g_sh - grad_i_recomb  # di/dvd
+        if breakdown_factor > 0:  # reverse bias is considered
+            brk_pwr_1 = np.power(brk_term, -breakdown_exp - 1)
+            brk_pwr_2 = np.power(brk_term, -breakdown_exp - 2)
+            brk_fctr = breakdown_factor * g_sh
+            grad_i_brk = brk_fctr * (brk_pwr + diode_voltage *
+                                     -breakdown_exp * brk_pwr_1)
+            grad2i_brk = brk_fctr * -breakdown_exp * (
+                2 * brk_pwr_1 + diode_voltage * (-breakdown_exp - 1) *
```
Comment (CameronTStark) [inline pvlib/singlediode.py:198]:
```suggestion
            grad2i_brk = (
                brk_fctr
                * -breakdown_exp
                * (2 * brk_pwr_1 + diode_voltage * (-breakdown_exp - 1) * brk_pwr_2)
            )

```

File: pvlib/singlediode.py [hunk 4]
```diff
@@ -150,21 +173,39 @@ def bishop88(diode_voltage, photocurrent, saturation_current,
...
     g_sh = 1.0 / resistance_shunt  # conductance
-    i = (photocurrent - saturation_current * np.expm1(v_star)
-         - diode_voltage * g_sh - i_recomb)
+    if breakdown_factor > 0:  # reverse bias is considered
+        brk_term = 1 - diode_voltage / breakdown_voltage
+        brk_pwr = np.power(brk_term, -breakdown_exp)
+        i_breakdown = breakdown_factor * diode_voltage * g_sh * brk_pwr
+    else:
+        i_breakdown = 0.
+    i = (photocurrent - saturation_current * np.expm1(v_star) -
```
Comment (CameronTStark) [inline pvlib/singlediode.py:182]:
```suggestion
    i = (
        photocurrent
        - saturation_current * np.expm1(v_star)
        - diode_voltage * g_sh
        - i_recomb
    )
```

File: pvlib/singlediode.py (hunk 2 above)
Comment (cwhanse) [reply pvlib/singlediode.py:222]: I agree, but would like to keep like arguments grouped together in general order of decreasing interest: required module parameters first, then optional thin-film parameters, reverse bias parameters third, numerical control last.

File: pvlib/singlediode.py [hunk 5]
```diff
@@ -150,21 +173,49 @@ def bishop88(diode_voltage, photocurrent, saturation_current,
...
+        i_breakdown = breakdown_factor * diode_voltage * g_sh * brk_pwr
+    else:
+        i_breakdown = 0.
+    i = (
+        photocurrent
+        - saturation_current * np.expm1(v_star)
+        - diode_voltage * g_sh
+        - i_recomb
+    )
+         diode_voltage * g_sh - i_recomb - i_breakdown)
```
Comment (wholmgren) [inline pvlib/singlediode.py]: E111/E113 needs to be fixed. not sure what's happening on this line anyways - bad merge? W503 above can be ignored.

File: pvlib/singlediode.py [hunk 6]
```diff
@@ -150,21 +173,49 @@ def bishop88(diode_voltage, photocurrent, saturation_current,
...
+            brk_pwr_2 = np.power(brk_term, -breakdown_exp - 2)
+            brk_fctr = breakdown_factor * g_sh
+            grad_i_brk = brk_fctr * (brk_pwr + diode_voltage *
+                                     -breakdown_exp * brk_pwr_1)
+            grad2i_brk = brk_fctr * -breakdown_exp * (
+            grad2i_brk = (
+                brk_fctr
+                * -breakdown_exp
+                * (2 * brk_pwr_1 + diode_voltage * (-breakdown_exp - 1) * brk_pwr_2)
+            )
```
Comment (wholmgren) [inline pvlib/singlediode.py]: I don't fully understand the stickler comments here. 

`-breakdown_exp` is used a lot. I recommend new variable for this. I also don't know the python operation binding rules well enough to immediately see that the math is going to be evaluated as intended.

File: pvlib/singlediode.py (hunk 2 above)
Comment (wholmgren) [reply pvlib/singlediode.py:228]: I think it's bad form to rely on the position of a keyword argument. Don't worry about it.

File: pvlib/singlediode.py [hunk 7]
```diff
@@ -351,9 +431,17 @@ def bishop88_mpp(photocurrent, saturation_current, resistance_series,
...
+    breakdown_factor : numeric, default 0
+        fraction of ohmic current involved in avalanche breakdown :math:`a`.
+        Default of 0 excludes the reverse bias term from the model. [unitless]
+    breakdown_voltage : numeric, default -5.5
+        reverse breakdown voltage of the photovoltaic junction :math:`V_{br}`
+        [V]
+    breakdown_exp : numeric, default 3.28
+        avalanche breakdown exponent :math:`m` [unitless]
+    method : str, default 'newton'
+        must be either ``'newton'`` or ``'brentq'``. If ``'breakdown_factor'``
```
Comment (wholmgren) [inline pvlib/singlediode.py:439]: Does `breakdown_factor` need `'` around it?

Comment (cwhanse) [reply pvlib/singlediode.py:439]: Replacing with 
``` 
   method : str, default 'newton'
        must be either ``'newton'`` or ``'brentq'``. If ``breakdown_factor``
        is not 0, ''method'' must be ``'newton'``.
```

**Comment by CameronTStark** [comment]
Thanks @cwhanse!

**Pull request merged**